src/api/venue.ts

```diff
@@ -34,3 +48,144 @@ export const setVenueLiveStatus = async ({
     })
     .finally(onFinish);
 };
+
+export interface FetchSovereignVenueOptions {
+  previouslyCheckedVenueIds?: readonly string[];
+  maxDepth?: number;
+}
+
+export interface FetchSovereignVenueReturn {
+  sovereignVenue: WithId<AnyVenue>;
+  checkedVenueIds: readonly string[];
+}
+
+export const fetchSovereignVenue = async (
+  venueId: string,
+  options?: FetchSovereignVenueOptions
+): Promise<FetchSovereignVenueReturn> => {
+  const { previouslyCheckedVenueIds = [], maxDepth } = options ?? {};
+
+  const venue = await fetchVenue(venueId);
+
+  if (!venue) throw new Error(`The '${venueId}' venue doesn't exist`);
+
+  if (!venue.parentId)
+    return {
+      sovereignVenue: venue,
+      checkedVenueIds: previouslyCheckedVenueIds,
+    };
+
+  if (previouslyCheckedVenueIds.includes(venueId))
+    throw new Error(
+      `Circular reference detected. '${venueId}' has already been checked`
+    );
+
+  if (maxDepth && maxDepth <= 0)
+    throw new Error("Maximum depth reached before finding the sovereignVenue.");
```

Comment:
What's the usecase for this? I think we always want to reach the sovereign venue no matter how far in the hierarchy it is

Reply:
Best practice patterns for recursive functions. It's optional and not currently used anyway, but I expect we may want to implement a 'sane upper limit' at some stage (eg. venues can't be nested more than 50 levels deep or something that most normal users are never going to hit, but that protects us/our code), at which point this would be used for it.

I'd almost go as far as to say that we should define that default (in `src/settings`) now, and make it the default value of this prop/option here.

Having this limit in place would also make it impossible to find yourself in infinite recursive loops, as it would 'bail out' once that limit is hit.

In frontend code like this it's slightly less important, if this was running in a backend function it'd be super important to prevent excess computation time/etc (be it by accident, or by malicious intent).

Reply:
OOS, will handle this in a followup PR.